symbolic-debuginfo/src/wasm.rs

```diff
@@ -0,0 +1,341 @@
+//! Support for WASM Objects (WebAssembly).
+use std::borrow::Cow;
+use std::fmt;
+
+use thiserror::Error;
+
+use symbolic_common::{Arch, AsSelf, CodeId, DebugId};
+
+use crate::base::*;
+use crate::dwarf::{Dwarf, DwarfDebugSession, DwarfError, DwarfSection, Endian};
+use crate::private::Parse;
+
+/// An error when dealing with [`WasmObject`](struct.WasmObject.html).
+#[non_exhaustive]
+#[derive(Debug, Error)]
+pub enum WasmError {
+    /// The module cannot be parsed.
+    #[error("invalid WASM file")]
+    BadObject,
+}
+
+/// Wasm object container (.wasm), used for executables and debug
+/// companions on web and wasi.
+///
+/// This can only parse binary wasm file and not wast files.
+pub struct WasmObject<'data> {
+    wasm_module: walrus::Module,
+    code_offset: u64,
+    data: &'data [u8],
+}
+
+impl<'data> WasmObject<'data> {
+    /// Tests whether the buffer could contain a WASM object.
+    pub fn test(data: &[u8]) -> bool {
+        data.starts_with(b"\x00asm")
+    }
+
+    /// Tries to parse a WASM from the given slice.
+    pub fn parse(data: &'data [u8]) -> Result<Self, WasmError> {
+        let wasm_module = match walrus::Module::from_buffer(data) {
+            Ok(module) => module,
+            Err(_) => return Err(WasmError::BadObject),
+        };
+
+        // we need to parse the file a second time to get the offset to the
+        // code section as walrus does not expose that yet.
+        let mut code_offset = 0;
+        for payload in wasmparser::Parser::new(0).parse_all(data) {
+            if let Ok(wasmparser::Payload::CodeSectionStart { range, .. }) = payload {
+                code_offset = range.start as u64;
+                break;
+            }
+        }
+
+        Ok(WasmObject {
+            wasm_module,
+            data,
+            code_offset,
+        })
+    }
+
+    /// The container file format, which is always `FileFormat::Wasm`.
+    pub fn file_format(&self) -> FileFormat {
+        FileFormat::Wasm
+    }
+
+    /// The code identifier of this object.
+    ///
+    /// Wasm does not yet provide code IDs.
+    pub fn code_id(&self) -> Option<CodeId> {
+        // see `debug_id`
+        for (_, section) in self.wasm_module.customs.iter() {
+            if section.name() == "build_id" {
+                return Some(CodeId::from_binary(&section.data(&Default::default())));
```

Comment:
can you extract the loop and returning the `section.data` into its own function, also, what does the `Default` expand to here? Its a bit hard to follow without actually jumping to the definition.